### DISCUSSION TITLE
AttributeRuler for `is_stop` gets "cached"

### DISCUSSION THREAD

**Discussion by maxtrem** [explosion/spaCy#7838] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

Hi, 
I was trying to add the word "order" as a stop word when it comes in the context "in order to".

I used the attribute ruler to make a rule for this however as soon as the attribute ruler is triggered all other tokens "order" are also marked as stop words, no matter in which context, see the example below.

**Code:**
```
import spacy, sys
nlp = spacy.load('en_core_web_sm')
ruler = nlp.get_pipe("attribute_ruler")

patterns = [
    {
        "patterns":[[{"LOWER": "in"}, {"LOWER": "order"}, {"LOWER": "to"}]],
        "attrs": {"IS_STOP": True},
        "index": 1,
    }
]

ruler.add_patterns(patterns)
print('Spacy Version:', spacy.__version__)
print('Python Version:', sys.version)

doc = nlp('in order')
print('doc:', doc)
for t in doc:
    print(f' - token: "{t}"  \t is_stop: {t.is_stop}')
print()
    
doc = nlp('order to')
print('doc:', doc)
for t in doc:
    print(f' - token: "{t}"  \t is_stop: {t.is_stop}')
print()

doc = nlp('in order to')
print('doc:', doc)
for t in doc:
    print(f' - token: "{t}"  \t is_stop: {t.is_stop}')
print()

doc = nlp('in order')
print('doc:', doc)
for t in doc:
    print(f' - token: "{t}"  \t is_stop: {t.is_stop}')
print()
    
doc = nlp('order to')
print('doc:', doc)
for t in doc:
    print(f' - token: "{t}"  \t is_stop: {t.is_stop}')
print()
```

**Output:**

```
Spacy Version: 3.0.5
Python Version: 3.8.2 (default, Mar 26 2020, 15:53:00) 
[GCC 7.3.0]
doc: in order
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: False

doc: order to
 - token: "order"  	 is_stop: False
 - token: "to"  	 is_stop: True

doc: in order to
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: True
 - token: "to"  	 is_stop: True

doc: in order
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: True   # should be False again

doc: order to
 - token: "order"  	 is_stop: True   # should be False again
 - token: "to"  	 is_stop: True
```

(I guess this is some kind of lookup and the attribute ruler might add "order" to the lookup once triggered?)

Is there some easy fix for this behaviour? If not, is there an alternative solution to archive this behaviour? 

Thanks!
/Max

## Your Environment

* Operating System: Ubuntu 20.04
* Python Version Used: 3.8.2 (default, Mar 26 2020, 15:53:00) 
[GCC 7.3.0]
* spaCy Version Used: 3.0.5

**Comment by maxtrem** [comment]
I was  actually able to solve (work around) the problem by adding another rule that generally sets the token `"order".is_stop` to `False`:

```
patterns = [
    {
        "patterns": [[{"LOWER": "order"}]],
        "attrs": {"IS_STOP": False},
    },
    {
        "patterns": [[{"LOWER": "in"}, {"LOWER": "order"}, {"LOWER": "to"}]],
        "attrs": {"IS_STOP": True},
        "index": 1,
    }
]
```

This of course is more inefficient than it needs to be, but it does the job. But it would probably be helpful mentioning this limitation somehow.

**Comment by adrianeboyd** [comment, marked as answer]
Hmm, I'm not sure that workaround is going to work?

`IS_STOP` is a `Lexeme` attribute, not a `Token` attribute, so there's just one lexeme entry for `order`, which is either a stop word or not.

(`NORM` is the only lexeme feature that's on both lexemes and tokens, so you can override `norm` for individual tokens.)

An alternative is to use a custom token extension to store whether individual tokens are stop words. It could back off to the lexeme attribute unless it's set.

**Reply by adrianeboyd** [reply]
~~I guess if each doc is processed immediately before the attribute ruler runs on the next doc this workaround would work, but it's not really setting `is_stop` for an individual token at any point.~~

It doesn't really work, since other occurrences of `order` in the same doc will also be affected.

**Reply by maxtrem** [reply]
Hej and thanks for your reply!
It seems to work though, with the changed rule set the output is as "expected":

```
patterns = [
    {
        "patterns": [[{"LOWER": "order"}]],
        "attrs": {"IS_STOP": False},
    },
    {
        "patterns": [[{"LOWER": "in"}, {"LOWER": "order"}, {"LOWER": "to"}]],
        "attrs": {"IS_STOP": True},
        "index": 1,
    }
]
```

**Output:**

```
Spacy Version: 3.0.5
Python Version: 3.8.2 (default, Mar 26 2020, 15:53:00) 
[GCC 7.3.0]
doc: in order
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: False

doc: order to
 - token: "order"  	 is_stop: False
 - token: "to"  	 is_stop: True

doc: in order to
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: True
 - token: "to"  	 is_stop: True

doc: in order
 - token: "in"  	 is_stop: True
 - token: "order"  	 is_stop: False

doc: order to
 - token: "order"  	 is_stop: False
 - token: "to"  	 is_stop: True
```

I guess if the rule that resets the "order" token lexeme entry comes first it will work and as far as I can tell for now this works, even for much larger amounts of data. So I guess the order of rules is kept internally.

I have thought about a custom extension, but this requires always some custom implementation (which could grow quite complex for more complex rules) while the AttributeRuler comes just with the right tools to also detect more complex patterns.

**Reply by maxtrem** [reply]
I just saw your second reply (I was first a bit puzzled when I suddenly could not post my reply :).
Yes, unfortunately for the same doc it does not work.

**Reply by adrianeboyd** [reply]
Sorry about that! Some of the behavior around converting to discussions is still a bit rough around the edges...

The attribute ruler can still be used to set custom extensions. It would look like custom extension matcher patterns, so something like this:

```python
Token.set_extension("custom_stop", default=False)

patterns = [
    {
        "patterns": [[{"LOWER": "in"}, {"LOWER": "order"}, {"LOWER": "to"}]],
        "attrs": {"_": {"custom_stop": True}},
        "index": 1,
    }
]
```

**Reply by maxtrem** [reply]
Is there some easy way to have a similar interface (like the AttributeRuler has) when using a custom attribute?
Previously I used token extensions mostly with some `if` expressions, but with this approach I would not come too far when having multiple expressions I want to match / add rules for. 
The interface of the AttributeRuler is kind of ideal as it allows you also to directly specify which token(s) you want to apply the attribute when having a multi-token match.

**Reply by maxtrem** [reply]
Thanks, that sounds great! I will try that (ignore my last answer, I just did not see your reply before posting it)

**Reply by maxtrem** [reply]
This works really great, thanks a lot! :)
(also really useful to know that custom extensions and the attribute ruler work together)
